staging/src/k8s.io/apiserver/pkg/storage/cacher/btree_store.go

```diff
@@ -0,0 +1,408 @@
+	}
+	storeElem, ok := obj.(*storeElement)
+	if !ok {
+		return fmt.Errorf("obj not a storeElement: %#v", obj)
+	}
+	_, err := s.deleteElem(storeElem)
+	return err
+}
+
+func (s *btreeStore) deleteElem(storeElem *storeElement) (interface{}, error) {
+	item := s.tree.Delete(storeElem)
+	if item == nil {
+		return nil, fmt.Errorf("obj does not exist")
+	}
+	return item, nil
+}
+
+func (s *btreeStore) List() []interface{} {
+	items := make([]interface{}, 0, s.tree.Len())
+	s.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(interface{}))
+		return true
+	})
+
+	return items
+}
+
+func (s *btreeStore) ListKeys() []string {
+	items := make([]string, 0, s.tree.Len())
+	s.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(*storeElement).Key)
+		return true
+	})
+
+	return items
+}
+
+func (s *btreeStore) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	storeElem, ok := obj.(*storeElement)
+	if !ok {
+		return nil, false, fmt.Errorf("obj is not a storeElement")
+	}
+	item = s.tree.Get(storeElem)
+	if item == nil {
+		return nil, false, nil
+	}
+
+	return item, true, nil
+}
+
+func (s *btreeStore) GetByKey(key string) (item interface{}, exists bool, err error) {
+	return s.getByKey(key)
+}
+
+func (s *btreeStore) Replace(objs []interface{}, _ string) error {
+	s.tree.Clear(false)
+	for _, obj := range objs {
+		storeElem, ok := obj.(*storeElement)
+		if !ok {
+			return fmt.Errorf("obj not a storeElement: %#v", obj)
+		}
+		s.addOrUpdateElem(storeElem)
+	}
+	return nil
+}
+
+// addOrUpdateLocked assumes a lock is held and is used for Add
+// and Update operations.
+func (s *btreeStore) addOrUpdateElem(storeElem *storeElement) *storeElement {
+	oldObj := s.tree.ReplaceOrInsert(storeElem)
+	if oldObj == nil {
+		return nil
+	}
+	return oldObj.(*storeElement)
+}
+
+func (s *btreeStore) getByKey(key string) (item interface{}, exists bool, err error) {
+	keyElement := &storeElement{Key: key}
+	item = s.tree.Get(keyElement)
+
+	return item, item != nil, nil
+}
+
+func (s *btreeStore) ListPrefix(prefix, continueKey string, limit int) ([]interface{}, bool) {
+	if limit < 0 {
+		return nil, false
+	}
+	if continueKey == "" {
+		continueKey = prefix
+	}
+	var result []interface{}
+	var hasMore bool
+
+	if limit == 0 {
+		s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+			elementKey := i.(*storeElement).Key
+			if !strings.HasPrefix(elementKey, prefix) {
+				return false
+			}
+			result = append(result, i.(interface{}))
+			return true
+		})
+		return result, hasMore
+	}
+
+	s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+		elementKey := i.(*storeElement).Key
+		if !strings.HasPrefix(elementKey, prefix) {
+			return false
+		}
+		// TODO: Might be worth to lookup one more item to provide more accurate HasMore.
+		if len(result) < limit {
+			result = append(result, i.(interface{}))
+		} else {
+			hasMore = true
+			return false
```

Comment:
Cannot, our integration tests would break. This is because it would make watch cache behavior inconsistent with etcd as it is basically extending lookup limit by 1.

Reply:
Update the comment? The way it is currently written suggests this is something that we *could* do but just didn't get around to.

What are the implications of this?  How do we make sure that any code that expects hasMore to be returned accurately is not impacted (now or in the future) by this change?

Reply:
>  Or is this actually compliant to the expected behavior of hasMore, just not optimal?

Yes, +1 lookup would break existing compliance test. I wrote the comment long time ago, thinking about it more, it's just +/- optimization. Not worth considering as most pagination limits should be 100+, so fact that we better handle exactly the 100 objects out of 100 page size is not very impactful. Will remove the comment. cc @wojtek-t for thoughts.